channeldb/payments.go

```diff
@@ -383,6 +383,139 @@ func fetchHtlcFailInfo(bucket *bbolt.Bucket) (*HTLCFailInfo, error) {
+	// fully completed. This means that pending payments, as well as failed
+	// payments will show up if this field is set to true.
+	IncludeIncomplete bool
+}
+
+// PaymentsQuerySlice contains the result of a query to the payments database.
+// It includes the original query, the set of payments that match the query,
+// and integers which represent the offset index of the first and last item
+// returned in the series of payments. These integers allow callers to resume
+// their query in the event that the query's response exceeds the max number of
+// returnable events.
+type PaymentsQuerySlice struct {
+	PaymentsQuery
+
+	// Payments is the set of payments returned from the database for the
+	// PaymentsQuery.
+	Payments []MPPayment
+
+	// FirstIndexOffset is the index of the first element in the set of
+	// returned MPPayments. Callers can use this to resume their query
+	// in the event that the slice has too many events to fit into a single
+	// response. The offset can be used to continue reverse pagination.
+	FirstIndexOffset uint64
+
+	// LastIndexOffset is the index of the last element in the set of
+	// returned MPPayments. Callers can use this to resume their query
+	// in the event that the slice has too many events to fit into a single
+	// response. The offset can be used to continue forward pagination.
+	LastIndexOffset uint64
+}
+
+// QueryPayments is a query to the payments database which is restricted
+// to a subset of payments by the payments query, containing an offset
+// index and a maximum number of returned payments.
+func (db *DB) QueryPayments(query PaymentsQuery) (PaymentsQuerySlice, error) {
+	resp := PaymentsQuerySlice{
+		PaymentsQuery: query,
+	}
+	allPayments, err := db.FetchPayments()
+	if err != nil {
+		return resp, err
+	}
+
+	if len(allPayments) == 0 {
+		return resp, nil
+	}
+
+	indexLimit := query.IndexOffset
+	// If the index limit is the default 0 value, we set our limit to our
+	// highest sequence number.
+	if indexLimit == 0 && query.Reversed {
+		indexLimit = allPayments[len(allPayments)-1].sequenceNum + 1
+	}
+
+	for i := range allPayments {
+		// If we have the max number of payments we want, exit.
+		if uint64(len(resp.Payments)) == query.MaxPayments {
+			break
+		}
+
+		index := i
+		if query.Reversed {
+			index = len(allPayments) - 1 - i
+		}
+
+		payment := allPayments[index]
+
+		// If we are paginating backwards, skip over all payments that
+		// have sequence numbers greater than or equal to the index
+		// offset. We skip payments with equal index because the
+		// offset is exclusive.
+		if query.Reversed && payment.sequenceNum >= indexLimit {
+			continue
+		}
+
+		// If we are paginating forwards, skip over all payments that
+		// have sequence numbers less than or equal to the index offset.
+		// We skip payments with equal indexes because the index offset
+		// is exclusive.
+		if !query.Reversed && payment.sequenceNum <= indexLimit {
+			continue
+		}
```

Comment:
Maybe it is clearer to group all code for forward and all code for reverse together in one if-then-else. But up to your preference.

Reply:
Yes, that's clearer. Regrouped the logic into reverse and forward directions.